store/sqlstore/team_store.go

```diff
@@ -372,45 +372,118 @@ func (s SqlTeamStore) GetByNames(names []string) ([]*model.Team, *model.AppError
 	return teams, nil
 }
 
+func (s SqlTeamStore) teamSearchQuery(term string, opts *model.TeamSearch, countQuery bool) sq.SelectBuilder {
+	var selectStr string
+	if countQuery {
+		selectStr = "count(*)"
+	} else {
+		selectStr = "*"
+	}
+
+	query := s.getQueryBuilder().
+		Select(selectStr).
+		From("Teams as t")
+
+	// Don't order or limit if getting count
+	if !countQuery {
+		query = query.OrderBy("t.DisplayName")
+
+		if opts.IsPaginated() {
+			query = query.Limit(uint64(*opts.PerPage)).Offset(uint64(*opts.Page * *opts.PerPage))
+		}
+	}
+
+	if len(term) > 0 {
+		term = sanitizeSearchTerm(term, "\\")
+		term = wildcardSearchTerm(term)
+
+		operatorKeyword := "ILIKE"
+		if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+			operatorKeyword = "LIKE"
+		}
+
+		query = query.Where(fmt.Sprintf("(Name %[1]s ? OR DisplayName %[1]s ?)", operatorKeyword), term, term)
+	}
+
+	var teamFilters sq.Sqlizer
+	var openInviteFilter sq.Sqlizer
+	if opts.AllowOpenInvite != nil {
+		if *opts.AllowOpenInvite {
+			openInviteFilter = sq.Eq{"AllowOpenInvite": true}
+		} else {
+			openInviteFilter = sq.And{
+				sq.Or{
+					sq.NotEq{"AllowOpenInvite": true},
+					sq.Eq{"AllowOpenInvite": nil},
+				},
+				sq.Or{
+					sq.NotEq{"GroupConstrained": true},
+					sq.Eq{"GroupConstrained": nil},
+				},
+			}
+		}
+
+		teamFilters = openInviteFilter
+	}
+
+	var groupConstrainedFilter sq.Sqlizer
+	if opts.GroupConstrained != nil {
+		if *opts.GroupConstrained {
+			groupConstrainedFilter = sq.Eq{"GroupConstrained": true}
+		} else {
+			groupConstrainedFilter = sq.Or{
+				sq.NotEq{"GroupConstrained": true},
+				sq.Eq{"GroupConstrained": nil},
+			}
+		}
+
+		if teamFilters == nil {
+			teamFilters = groupConstrainedFilter
+		} else {
+			teamFilters = sq.Or{teamFilters, groupConstrainedFilter}
+		}
+	}
```

Comment:
@mkraft @hahmadia I was able to resolve this in a much cleaner way now by excluding group constrained teams when `AllowOpenInvite = false` essentially treating `Private` teams as a completely different entity than `GroupConstrained` teams. If the user wants to get all `Private` teams including `GroupConstrained` then they have to simply set `AllowOpenInvite = false` and `GroupConstrained = true`